File: src/dotnet/Core/Application/Services/SeniorityMatcher.cs
using Rag.Candidates.Core.Domain.Configuration;
using Rag.Candidates.Core.Domain.Enums;

namespace Rag.Candidates.Core.Application.Services;

public sealed class SeniorityMatcher
{
    private readonly QueryParsingConfig _config;
    private readonly Dictionary<string, SeniorityLevel> _tokenMap;

    public SeniorityMatcher(QueryParsingConfig config)
    {
        _config = config;
        _tokenMap = BuildTokenMap();
    }

    private Dictionary<string, SeniorityLevel> BuildTokenMap()
    {
        var mapping = new Dictionary<string, SeniorityLevel>();
        
        foreach (var (level, tokens) in _config.SeniorityTokens)
        {
            foreach (var token in tokens)
            {
                mapping[token.ToLowerInvariant()] = level;
            }
        }
        
        return mapping;
    }

    public SeniorityLevel? Match(string queryText)
    {
        var queryLower = queryText.ToLowerInvariant();
        
        foreach (var (token, level) in _tokenMap)
        {
            if (queryLower.Contains(token))
            {
                return level;
            }
        }
        
        return null;
    }
}

File: src/python/core/application/services/seniority_matcher.py
from typing import Optional, Dict
from ...domain.enums.seniority_level import SeniorityLevel
from ...domain.configuration.query_parsing_config import QueryParsingConfig


class SeniorityMatcher:
    def __init__(self, config: QueryParsingConfig):
        self._config = config
        self._token_map = self._build_token_map()
    
    def _build_token_map(self) -> Dict[str, SeniorityLevel]:
        mapping = {}
        for level, tokens in self._config.seniority_tokens.items():
            for token in tokens:
                mapping[token.lower()] = level
        return mapping
    
    def match(self, query_text: str) -> Optional[SeniorityLevel]:
        query_lower = query_text.lower()
        
        for token, level in self._token_map.items():
            if token in query_lower:
                return level
        
        return None
